File: updatewithgithub/src/main/java/velitasali/updatewithgithub/GithubUpdater.java
package velitasali.updatewithgithub;

import android.app.DownloadManager;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.os.Looper;
import android.support.v7.app.AlertDialog;
import android.view.WindowManager;
import android.widget.Toast;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.File;
import java.net.URI;

/**
 * Created by: veli
 * Date: 11/13/16 8:25 AM
 */

public class GithubUpdater
{
	private Context mContext;
	private String mRepo;
	private int mThemeRes;

	public GithubUpdater(Context context, String repo, int themeRes)
	{
		this.mContext = context;
		this.mRepo = repo;
		this.mThemeRes = themeRes;
	}

	public void checkForUpdates()
	{
		new Thread()
		{
			@Override
			public void run()
			{
				super.run();

				Looper.prepare();

				try
				{
					Toast.makeText(mContext, R.string.uwg_check_for_updates_ongoing, Toast.LENGTH_LONG).show();

					mContext.setTheme(mThemeRes);

					RemoteServer server = new RemoteServer(mRepo);
					String result = server.connect(null, null);

					PackageInfo packInfo = mContext.getPackageManager().getPackageInfo(mContext.getApplicationInfo().packageName, 0);

					String appVersionName = packInfo.versionName;
					int appVersionCode = packInfo.versionCode;
					final String applicationName = getAppLabel(mContext);

					JSONArray releases = new JSONArray(result);

					if (releases.length() > 0)
					{
						JSONObject lastRelease = releases.getJSONObject(0);

						final String availableVersion = lastRelease.getString("tag_name");

						ComparableVersion comLast = new ComparableVersion(availableVersion);
						ComparableVersion comCurr = new ComparableVersion(appVersionName);

						final File updateFile = new File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS) + "/" + applicationName + " v" + availableVersion + ".apk");

						if (comLast.compareTo(comCurr) > 0)
						{
							if (lastRelease.has("assets"))
							{
								JSONArray releaseAssets = lastRelease.getJSONArray("assets");

								if (releaseAssets.length() > 0)
								{
									final JSONObject firstAsset = releaseAssets.getJSONObject(0);
									final String downloadURL = firstAsset.getString("browser_download_url");

									int negativeButtonLabel;
									int positiveButtonLabel;
									final String updateBody;
									DialogInterface.OnClickListener positiveButtonListener;

									DialogInterface.OnClickListener downloadButtonListener = new DialogInterface.OnClickListener()
									{
										@Override
										public void onClick(DialogInterface dialog, int which)
										{
											if (updateFile.isFile())
												updateFile.delete();

											DownloadManager manager = (DownloadManager) mContext.getSystemService(Context.DOWNLOAD_SERVICE);
											DownloadManager.Request downloadRequest = new DownloadManager.Request(Uri.parse(downloadURL));

											downloadRequest.setTitle(mContext.getString(R.string.uwg_downloading_update_title, applicationName, availableVersion));
											downloadRequest.setDescription(mContext.getString(R.string.uwg_downloading_update_description, applicationName));
											downloadRequest.setDestinationInExternalPublicDir(Environment.DIRECTORY_DOWNLOADS, applicationName + " v" + availableVersion + ".apk");
											downloadRequest.setMimeType("application/vnd.android.package-archive");
											if (Build.VERSION.SDK_INT >= 11)
												downloadRequest.setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED);

											manager.enqueue(downloadRequest);
										}
									};

									DialogInterface.OnClickListener updateButtonListener = new DialogInterface.OnClickListener()
									{
										@Override
										public void onClick(DialogInterface dialog, int which)
										{
											mContext.startActivity(new Intent(Intent.ACTION_VIEW).setDataAndType(Uri.fromFile(updateFile), "application/vnd.android.package-archive"));
										}
									};

									if (updateFile.isFile())
									{
										if (firstAsset.getLong("size") == updateFile.length())
										{
											negativeButtonLabel = R.string.uwg_later;
											positiveButtonLabel = R.string.uwg_update_now;
											updateBody = mContext.getString(R.string.uwg_update_now_info);
											positiveButtonListener = updateButtonListener;
										}
										else
										{
											negativeButtonLabel = R.string.uwg_cancel;
											positiveButtonLabel = R.string.uwg_download_anyway;
											updateBody = mContext.getString(R.string.uwg_update_file_different_info);
											positiveButtonListener = downloadButtonListener;
										}
									}
									else
									{
										negativeButtonLabel = R.string.uwg_later;
										positiveButtonLabel = R.string.uwg_download_now;
										updateBody = String.format(mContext.getString(R.string.uwg_update_body), appVersionName, availableVersion, lastRelease.getString("published_at"), lastRelease.getString("body"));

										positiveButtonListener = downloadButtonListener;
									}

									AlertDialog.Builder builder = new AlertDialog.Builder(mContext);

									builder.setTitle(R.string.uwg_update_available);
									builder.setMessage(updateBody);

									builder.setNegativeButton(negativeButtonLabel, null);
									builder.setPositiveButton(positiveButtonLabel, positiveButtonListener);

									AlertDialog dialog = builder.create();

									dialog.getWindow().setType(WindowManager.LayoutParams.TYPE_SYSTEM_ALERT);
									dialog.show();
								}
							}
							else
								Toast.makeText(mContext, R.string.uwg_no_update_available, Toast.LENGTH_LONG).show();
						}
						else
							Toast.makeText(mContext, R.string.uwg_currently_latest_version_info, Toast.LENGTH_LONG).show();
					}
				} catch (Exception e) {
					Toast.makeText(mContext, R.string.uwg_version_check_error, Toast.LENGTH_LONG).show();
				}
				finally
				{
					Looper.loop();
				}
			}
		}.start();
	}


	public String getAppLabel(Context context) {
		PackageManager packageManager = context.getPackageManager();
		ApplicationInfo applicationInfo = null;
		try {
			applicationInfo = packageManager.getApplicationInfo(context.getApplicationInfo().packageName, 0);
		} catch (final PackageManager.NameNotFoundException e) {
		}
		return (String) (applicationInfo != null ? packageManager.getApplicationLabel(applicationInfo) : "Unknown");
	}
}

File: core/src/main/java/com/genonbeta/android/updatewithgithub/GitHubUpdater.kt
/*
 * Copyright (C) 2020 Veli Tasalı
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package com.genonbeta.android.updatewithgithub

import android.app.Activity
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.net.Uri
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.core.content.ContextCompat.startActivity
import com.genonbeta.android.framework.R
import org.apache.maven.artifact.versioning.ComparableVersion
import org.json.JSONArray


/**
 * Created by: veli
 * Date: 11/13/16 8:25 AM
 */

class GitHubUpdater(
    private val context: Context,
    private val repo: String,
    private val includePreReleases: Boolean,
) {
    fun checkForUpdates(popupDialog: Boolean, listener: OnInfoAvailableListener?) {
        if (popupDialog) {
            Toast.makeText(context, R.string.genfw_uwg_check_for_updates_ongoing, Toast.LENGTH_LONG).show()
        }

        val thread = Thread {
            try {
                Log.d(TAG, "Checking updates")

                val server = RemoteServer(repo)
                val result = server.connect(null, null)

                Log.d(TAG, "Server connected")

                val packageInfo = context.packageManager.getPackageInfo(context.applicationInfo.packageName, 0)
                val versionName = packageInfo.versionName
                val applicationName = getAppLabel()
                val releases = JSONArray(result)

                if (releases.length() > 0) {
                    Log.d(TAG, "Reading releases: (total) " + releases.length())

                    for (iterator in 0 until releases.length()) {
                        val releaseObject = releases.getJSONObject(iterator)
                        val isPreRelease = releaseObject.getBoolean("prerelease")
                        if (isPreRelease && !includePreReleases) continue
                        val updateVersion = releaseObject.getString("tag_name")
                        val updateTitle = releaseObject.getString("name")
                        val updateDate = releaseObject.getString("published_at")
                        val updateBody = releaseObject.getString("body")
                        val pageUrl = releaseObject.getString("html_url")
                        val comparableLatest = ComparableVersion(updateVersion)
                        val comparableCurrent = ComparableVersion(versionName)
                        val isNew = comparableLatest > comparableCurrent

                        if (context is Activity && !context.isFinishing) context.runOnUiThread {
                            listener?.onInfoAvailable(isNew, updateVersion, updateTitle, updateBody, updateDate)
                        } else
                            listener?.onInfoAvailable(isNew, updateVersion, updateTitle, updateBody, updateDate)

                        if (!popupDialog || context !is Activity) {
                            Log.d(TAG, "Skipping showing dialogs: requested=$popupDialog")
                        } else if (isNew) {
                            Log.d(TAG, "New version found: $updateVersion")

                            if (releaseObject.has("assets")) {
                                Log.d(TAG, "Reading assets")
                                val releaseAssets = releaseObject.getJSONArray("assets")

                                if (releaseAssets.length() > 0) {
                                    Log.d(TAG, "Assets is cached: (total) " + releaseAssets.length())

                                    val visitPage = DialogInterface.OnClickListener { _, _ ->
                                        startActivity(
                                            context, Intent(Intent.ACTION_VIEW, Uri.parse(pageUrl)),
                                            null
                                        )
                                    }

                                    context.runOnUiThread {
                                        AlertDialog.Builder(context)
                                            .setTitle(R.string.genfw_uwg_update_available)
                                            .setMessage(
                                                String.format(
                                                    context.getString(R.string.genfw_uwg_update_body),
                                                    versionName,
                                                    updateVersion,
                                                    updateDate,
                                                    updateBody
                                                )
                                            )
                                            .setPositiveButton(R.string.genfw_uwg_visit_page, visitPage)
                                            .setNegativeButton(android.R.string.cancel, null)
                                            .show()
                                    }

                                } else Log.d(TAG, "No downloadable file is provided")
                            } else context.runOnUiThread {
                                Toast.makeText(context, R.string.genfw_uwg_no_update, Toast.LENGTH_LONG).show()
                            }
                        } else context.runOnUiThread {
                            Toast.makeText(context, R.string.genfw_uwg_up_to_date, Toast.LENGTH_LONG).show()
                        }
                        break
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()

                if (popupDialog && context is Activity) context.runOnUiThread {
                    Toast.makeText(context, R.string.genfw_uwg_version_check_error, Toast.LENGTH_LONG).show()
                }
            }
        }

        thread.start()
    }

    private fun getAppLabel(): String? {
        val packageManager = context.packageManager

        try {
            packageManager.getApplicationLabel(
                packageManager.getApplicationInfo(context.applicationInfo.packageName, 0)
            )
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }
        return null
    }

    fun isNewVersion(versionName: String): Boolean {
        return isNewVersion(context, versionName)
    }

    interface OnInfoAvailableListener {
        fun onInfoAvailable(
            newVersion: Boolean,
            versionName: String?,
            title: String?,
            description: String?,
            releaseDate: String?,
        )
    }

    companion object {
        val TAG: String = GitHubUpdater::class.java.simpleName

        fun isNewVersion(context: Context, versionName: String): Boolean {
            try {
                val packageInfo: PackageInfo =
                    context.packageManager.getPackageInfo(context.applicationInfo.packageName, 0)
                val comparableGiven = ComparableVersion(versionName)
                val comparableCurrent = ComparableVersion(packageInfo.versionName)

                return comparableGiven > comparableCurrent
            } catch (e: PackageManager.NameNotFoundException) {
                e.printStackTrace()
            }
            return false
        }
    }
}